flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotRequestBulkCheckerImplTest.java

```diff
@@ -0,0 +1,257 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import static org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlotTestUtils.createPhysicalSlot;
+import static org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlotTestUtils.occupyPhysicalSlot;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link PhysicalSlotRequestBulkCheckerImpl}.
+ */
+public class PhysicalSlotRequestBulkCheckerImplTest extends TestLogger {
+
+	private static final Time TIMEOUT = Time.milliseconds(100L);
+
+	private static ScheduledExecutorService singleThreadScheduledExecutorService;
+
+	private static ComponentMainThreadExecutor mainThreadExecutor;
+
+	private final ManualClock clock = new ManualClock();
+
+	private PhysicalSlotRequestBulkCheckerImpl bulkChecker;
+
+	private Set<PhysicalSlot> slots;
+
+	private Supplier<Set<SlotInfo>> slotsRetriever;
+
+	@BeforeClass
+	public static void setupClass() {
+		singleThreadScheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+		mainThreadExecutor = ComponentMainThreadExecutorServiceAdapter.forSingleThreadExecutor(singleThreadScheduledExecutorService);
+	}
+
+	@AfterClass
+	public static void teardownClass() {
+		if (singleThreadScheduledExecutorService != null) {
+			singleThreadScheduledExecutorService.shutdownNow();
+		}
+	}
+
+	@Before
+	public void setup() throws Exception {
+		slots = new HashSet<>();
+		slotsRetriever = () -> new HashSet<>(slots);
+		bulkChecker = new PhysicalSlotRequestBulkCheckerImpl(slotsRetriever, clock);
+		bulkChecker.start(mainThreadExecutor);
+	}
+
+	@Test
+	public void testPendingBulkIsNotCancelled() throws InterruptedException, ExecutionException {
+		final CompletableFuture<SlotRequestId> cancellationFuture = new CompletableFuture<>();
+		final PhysicalSlotRequestBulk bulk = createPhysicalSlotRequestBulkWithCancellationFuture(cancellationFuture, new SlotRequestId());
+		bulkChecker.schedulePendingRequestBulkTimeoutCheck(bulk, TIMEOUT);
+		checkNotCancelledAfter(cancellationFuture, 2 * TIMEOUT.toMilliseconds());
+	}
+
+	@Test
+	public void testFulfilledBulkIsNotCancelled() throws InterruptedException, ExecutionException {
+		final CompletableFuture<SlotRequestId> cancellationFuture = new CompletableFuture<>();
+		final PhysicalSlotRequestBulk bulk = createPhysicalSlotRequestBulkWithCancellationFuture(cancellationFuture, new SlotRequestId());
+		bulkChecker.schedulePendingRequestBulkTimeoutCheck(bulk, TIMEOUT);
+		checkNotCancelledAfter(cancellationFuture, 2  * TIMEOUT.toMilliseconds());
+	}
+
+	private static void checkNotCancelledAfter(
+			CompletableFuture<?> cancellationFuture, long milli) throws ExecutionException, InterruptedException {
+		mainThreadExecutor.schedule(() -> {}, milli, TimeUnit.MILLISECONDS).get();
+		try {
+			cancellationFuture.get(milli, TimeUnit.MILLISECONDS);
```

Comment:
Maybe `checkThat(cancellationFuture.isDone(), is(false))`? Otherwise it will actually wait for `2 * milli` milliseconds in this method.

Reply:
True, I can add it. Although we have to wait in usual case.